src/setindex.jl

```diff
@@ -20,3 +20,9 @@ Base.@propagate_inbounds function setindex(d0::AbstractDict, v, k)
     d[k] = v
     return d
 end
+
+Base.@propagate_inbounds function setindex(ref::Base.RefValue{Tx}, val::Ty) where {Tx, Ty}
+    T = promote_type(Tx, Ty)
+    new_ref = Base.RefValue{T}(convert(T, val))
+    return new_ref
+end
```

Comment:
Why the `Base.@propagate_inbounds`? Also I think I would prefer not to promote here. I think just 
```julia
setindex(::Base.RefValue, val) = Ref(val)
```
is fine. Promotion can quickly lead to eltype `Any`.

Reply:
For other containers we do promote in `setindex`. However these containers typically have multiple elements and so the new container must contain both old elements and the new one added. A `Ref` however has always only a single element, so there is less need to ensure the old element can still be contained.

Reply:
I think the function should be force inlined, so it doesn't allocate if the original Ref was stack allocated, though `Base.@propagate_inbounds` isn't required since there is no indexing. Also, I am sure by default it'll get inlined by the compiler so I don't know if the explicitness is required.

Regarding promotion, you're right, there's no need for promotion or conversion.